sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java

```diff
@@ -3,32 +3,64 @@
 
 package com.azure.cosmos.implementation;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import javax.crypto.Mac;
 import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
 
 class MacPool {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MacPool.class);
     final Mac macInstance;
     final ConcurrentLinkedQueue<Mac> pool;
+    final AtomicBoolean isMacInstanceCloneable = new AtomicBoolean(true);
+    final Supplier<Mac> macProvider;
 
-    public MacPool(Mac rootMac) {
-        if (rootMac == null) {
-            throw new IllegalArgumentException("rootMac");
+    public MacPool(Supplier<Mac> macProvider) {
+        if (macProvider == null) {
+            throw new IllegalArgumentException("macProvider");
         }
 
-        this.macInstance = rootMac;
+        this.macProvider = macProvider;
+        this.macInstance = macProvider.get();
         this.pool = new ConcurrentLinkedQueue<>();
     }
 
     public ReUsableMac take() {
-        try {
-            Mac cachedInstance = pool.poll();
-            if (cachedInstance == null) {
-                cachedInstance = (Mac) this.macInstance.clone();
-            }
+        Mac cachedInstance = pool.poll();
+        if (cachedInstance == null) {
+            cachedInstance = this.createNewMac();
+        }
 
-            return new ReUsableMac(cachedInstance, this);
+        return new ReUsableMac(cachedInstance, this);
+    }
+
+    private Mac createNewMac() {
+        if (!this.isMacInstanceCloneable.get()) {
+            return this.macProvider.get();
+        }
+
+        // Cloning teh rootMac is the fastest option
+        // But while Mac implements Cloneable interface not every MacSpi is required to support cloning
+        // If it doesn't  - bad for performance - but functionally correct and necessary
+        // - see https://docs.oracle.com/javase/7/docs/technotes/guides/security/crypto/HowToImplAProvider.html#Step11
+        // - see https://docs.oracle.com/en/java/javase/18/docs/api/java.base/javax/crypto/Mac.html#clone()
+        try {
+            return (Mac) this.macInstance.clone();
         } catch (CloneNotSupportedException e) {
-            throw new IllegalStateException(e);
+
+            LOGGER.warn(
```

Comment:
Do we need this warning log?
I am concerned that customers might see this log and blame for perf issues, while at the same time not sure on how much control they have got on this?